File: solutions/0144_Binary_Tree_Preorder_Traversal/iter_with_stack.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        stack = [root]
        ret = []
        while stack:
            node = stack.pop()
            ret += [node.val]
            if node.right:
                stack += [node.right]
            if node.left:
                stack += [node.left]
        return ret

File: solutions/0144_Binary_Tree_Preorder_Traversal/iter_with_stack.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> ret = new ArrayList<>();
        if (root == null) return ret;
        Stack<TreeNode> rChilds = new Stack<>();
        TreeNode cur = root; // always keep left child
        while (cur != null) {
            ret.add(cur.val);
            if (cur.right != null) {
                rChilds.push(cur.right);
            }
            cur = cur.left; // recur on left child
            if (cur == null && rChilds.size() > 0) {
                cur = rChilds.pop();
            }
        }
        return ret;
    }
}
